### PR TITLE
[RFR] feat: add jsonapi format

### PR DESCRIPTION
**Pull request by Simperfit**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR | `@todo` |

Add the jsonApi formats to the supported format
- [x] Add test behat
- [x] Validate schema using [jsonApi-validator](https://www.npmjs.com/package/jsonapi-validator)

### CODE REVIEW

**Comment by Simperfit** [comment]
@dunglas styleci is bugging again

File: features/bootstrap/JsonApiContext.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        }
+        file_put_contents(dirname(__FILE__).'/response.json', $content);
+    }
+
+    /**
+     * @Then I valide it with jsonapi-validator
+     */
+    public function iValideItWithJsonapiValidator()
+    {
+        return 'response.json is valid JSON API.' === exec(sprintf('cd %s && jsonapi-validator -f response.json', dirname(__FILE__)));
```
Comment (soyuka) [inline features/bootstrap/JsonApiContext.php]: Can't we test through a simple schema instead of adding a new software and executing a system command?

Comment (Simperfit) [inline features/bootstrap/JsonApiContext.php]: I understand the point but this command contains a schema automatically updated, do we have do to so ?

Comment (meyerbaptiste) [inline features/bootstrap/JsonApiContext.php]: Or maybe you can load the JSON Schema from http://jsonapi.org/schema and then use the `\Sanpi\Behatch\Json\JsonInspector::validate()` method.

Comment (Simperfit) [inline features/bootstrap/JsonApiContext.php]: @meyerbaptiste Good point :)

Comment (Simperfit) [inline features/bootstrap/JsonApiContext.php]: @meyerbaptiste But that means that we need an internet connection to make the test works. @soyuka @dunglas WDYT ?

Comment (dunglas) [inline features/bootstrap/JsonApiContext.php]: You can copy the JSONAPI's schema directly in our test folder.

File: src/Hal/Serializer/CollectionNormalizer.php [hunk 2]
```diff
@@ -55,6 +55,9 @@ public function supportsNormalization($data, $format = null)
     public function normalize($object, $format = null, array $context = [])
     {
         $data = [];
+        $currentPage = 1;
```
Comment (dunglas) [inline src/Hal/Serializer/CollectionNormalizer.php]: Why those change? HAL support should not be modifier in this PR.

File: src/JsonApi/Serializer/EntrypointNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = [])
+    {
+        $entrypoint = ['links' => ['self' => $this->urlGenerator->generate('api_entrypoint')]];
+
+        foreach ($object->getResourceNameCollection() as $resourceClass) {
+            $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
+
+            if (empty($resourceMetadata->getCollectionOperations())) {
```
Comment (dunglas) [inline src/JsonApi/Serializer/EntrypointNormalizer.php]: `empty` can be removed.

File: src/JsonApi/Serializer/ItemNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    {
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function denormalize($data, $class, $format = null, array $context = [])
+    {
+        throw new RuntimeException(sprintf('%s is a read-only format.', self::FORMAT));
```
Comment (dunglas) [inline src/JsonApi/Serializer/ItemNormalizer.php]: Obviously it's not the case: http://jsonapi.org/format/#crud

**Comment by Simperfit** [comment]
Typical output 

```
{
  "links": {
    "self": "/books.jsonapi?page=1",
    "first": "/books.jsonapi?page=1",
    "last": "/books.jsonapi?page=2",
    "next": "/books.jsonapi?page=2"
  },
  "data": [
    {
      "type": "Book",
      "id": 1,
      "attributes": {
        "id": 1,
        "name": "Darko"
      }
    },
    {
      "type": "Book",
      "id": 2,
      "attributes": {
        "id": 2,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 3,
      "attributes": {
        "id": 3,
        "name": "fe"
      }
    },
    {
      "type": "Book",
      "id": 4,
      "attributes": {
        "id": 4,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 5,
      "attributes": {
        "id": 5,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 6,
      "attributes": {
        "id": 6,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 7,
      "attributes": {
        "id": 7,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 8,
      "attributes": {
        "id": 8,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 9,
      "attributes": {
        "id": 9,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 10,
      "attributes": {
        "id": 10,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 11,
      "attributes": {
        "id": 11,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 12,
      "attributes": {
        "id": 12,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 13,
      "attributes": {
        "id": 13,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 14,
      "attributes": {
        "id": 14,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 15,
      "attributes": {
        "id": 15,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 16,
      "attributes": {
        "id": 16,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 17,
      "attributes": {
        "id": 17,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 18,
      "attributes": {
        "id": 18,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 19,
      "attributes": {
        "id": 19,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 20,
      "attributes": {
        "id": 20,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 21,
      "attributes": {
        "id": 21,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 22,
      "attributes": {
        "id": 22,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 23,
      "attributes": {
        "id": 23,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 24,
      "attributes": {
        "id": 24,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 25,
      "attributes": {
        "id": 25,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 26,
      "attributes": {
        "id": 26,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 27,
      "attributes": {
        "id": 27,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 28,
      "attributes": {
        "id": 28,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 29,
      "attributes": {
        "id": 29,
        "name": "Ember Hamster"
      }
    },
    {
      "type": "Book",
      "id": 30,
      "attributes": {
        "id": 30,
        "name": "Ember Hamster"
      }
    }
  ],
  "meta": {
    "totalItems": 35,
    "itemsPerPage": 30
  }
}

```

Is the pagination wrong ?

**Comment by Simperfit** [comment]
closing in favor of #1036

**Pull request closed without merging**